### PR TITLE
Add changelog for v0.4.2

### PR DESCRIPTION
**Pull request by lgierth**
Everybody PTAL -- did I overlook anything?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,63 @@
...
+
+This is a patch release which fixes perfomance and networking bugs in go-libp2p,
+You should see improvements in CPU and RAM usage, as well as speed of object lookups.
+There are also a few other nice improvements.
+
+* Notable Fixes
+  * Set a deadline for dialing attempts. This prevents a node from accumulating
+    failed connections. (@whyrusleeping)
+  * Avoid unneccessary string/byte conversions in go-multihash. (@whyrusleeping)
+  * Fix a deadlock around the yamux stream muxer. (@whyrusleeping)
```
Comment (jbenet) [inline CHANGELOG.md]: may be good to start linking to the various issues, **shrug**

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,63 @@
...
+  * Fix a bug that left channels open, causing hangs. (@whyrusleeping)
+  * Fix a bug around yamux which caused connection hangs. (@whyrusleeping)
+
+* Enhancements
+  * Add NetBSD support. (@erde74)
+  * Set Cache-Control: immutable on /ipfs responses. (@kpcyrd)
+  * Have `ipfs init` optionally accept a default configuration from stdin. (@sivachandran)
+  * Add `ipfs log ls` command for listing logging subsystems. (@hsanjuan)
```
Comment (jbenet) [inline CHANGELOG.md]: A+ i wanted that a long time ago, thanks @hsanjuan

File: CHANGELOG.md (hunk 1 above)
Comment (lgierth) [inline CHANGELOG.md]: The entries here often don't match one-to-one with a PR -- but I agree that for the interesting changes it'd be pretty useful.

Next time? :) Or I can give it another go tomorrow.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,63 @@
...
+  * Allow bitswap to read multiple messages per stream. (@whyrusleeping)
+  * Remove `make toolkit_upgrade` step. (@chriscool)
+
+* Documentation
+  * Add a debug-guidelines document. (@richardlitt)
+  * Update the contribute document. (@richardlitt)
+  * Fix documentation of all `ipfs` commands. (@richardlitt)
```
Comment (RichardLitt) [inline CHANGELOG.md]: Should be more detailed than this.

Comment (RichardLitt) [inline CHANGELOG.md]: How about: `Improved default documentation and logic for most `ipfs` command options`

Comment (lgierth) [inline CHANGELOG.md]: Ok cool, suggestions? :)

Comment (RichardLitt) [inline CHANGELOG.md]: Wait, what is this meant to be for?

Comment (lgierth) [inline CHANGELOG.md]: Maybe it should be split into "Fix default values of many `ipfs` commands" and "Improve commands documentation"?

Comment (RichardLitt) [inline CHANGELOG.md]: But we cover that below. What were you referring to in this comment?

**Comment by whyrusleeping** [comment]
LGTM. I think its time to hop on the release train!

File: CHANGELOG.md (hunk 3 above)
Comment (RichardLitt) [inline CHANGELOG.md]: Resolved.

**Comment by lgierth** [comment]
This is good to go

**Pull request merged**